contracts/deployables/counter-sign/CounterSignV1.sol

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: AGPL-3.0
+pragma solidity ^0.8.30;
+
+import {ICounterSignV1} from "../../interfaces/decent/deployables/ICounterSignV1.sol";
+import {Initializable} from "@openzeppelin/contracts-upgradeable/proxy/utils/Initializable.sol";
+
+/*
+Design discussion: https://docs.google.com/document/d/1xp8w25O7CnsMf3cVPX_AZ3XT0k6DaAaWWl50p-VCj9A/edit?usp=sharing
+*/
+
+contract CounterSignV1 is Initializable, ICounterSignV1 {
+    string private agreementUrl;
+    uint256 private expirationTimestamp;
+    bool private executed;
+
+    address[] private signers;
+    mapping(address => string) private entityNames;
+    mapping(address => string) private signerPersonalNames;
+    mapping(address => bool) public signedAddresses;
+
+    Transaction[] private finalAgreementTransactions;
+
+    modifier onlyBeforeExpiry() {
+        require(block.timestamp < expirationTimestamp, "Agreement expired");
+        _;
+    }
+
+    function initialize(bytes memory params) external initializer {
+        (
+            address[] memory _signers,
+            string[] memory _entityNames,
+            uint256 _expiration,
+            string memory _agreementUrl,
+            Transaction[] memory _finalTxs
+        ) = abi.decode(
+                params,
+                (address[], string[], uint256, string, Transaction[])
+            );
+
+        require(
+            _signers.length == _entityNames.length,
+            "Signers and names mismatch"
+        );
```

Comment:
nit: use proper solidity `error` types instead of `require`.

```sol
error InputMismatch();

function initialize(...) {
  ...
  if (_signers.length != _entityNames.length) {
    revert InputMismatch();
  }
}
```